### ISSUE TITLE
`cabal.project` `source-repository-package` might be ignored

### ISSUE THREAD

**Issue by peterbecich** [srid/haskell-flake#212]
I appreciate `haskell-flake`'s support for `cabal.project`. I think it might be ignoring `source-repository-package`: https://cabal.readthedocs.io/en/stable/cabal-project.html#specifying-packages-from-remote-version-control-locations

`haskell-multi-nix` with example:
https://github.com/peterbecich/haskell-multi-nix/tree/source-repository-package-example

https://github.com/peterbecich/haskell-multi-nix/blob/source-repository-package-example/cabal.project

This should attempt to pull `haskell/text` bogus commit `123`
`cabal.project`
```
packages:
    ./foo
    ./bar

source-repository-package
  type: git
  location: https://github.com/haskell/text.git
  tag: 123
  -- tag: 73620de89d43ee50de2d15b7bc0843bf6d6e9b9a
```
`cabal build all` fails as it should. However, `nix build` succeeds; I think it is getting `text` from NixPkgs as usual.

Cabal pulls the real commit `73620de` correctly. `nix build` doesn't indicate if it is pulling this commit or not.

---------------------

I know the `text` dependency could be specified in `flake.nix`: https://community.flake.parts/haskell-flake/dependency/#path

Thank you

**Comment by shivaraj-bh** [comment]
`haskell-flake` only parses [packages from cabal.project](https://github.com/srid/haskell-flake/blob/e8ff18d1352bdfb0240fb721e7dbe2dbab2ec3e8/nix/haskell-parsers/parser.nix#L17-L31), nothing else.

**Comment by srid** [comment]
Yea, this is not supported, but we should document it clearly. You want to specify those source dependencies are flake inputs: https://community.flake.parts/haskell-flake/dependency#path

**Comment by peterbecich** [comment]
Thanks, https://community.flake.parts/haskell-flake/dependency#path works great.

If the `source-repository-package` is kept in `cabal.project`, it appears to cause an error in `haskell-flake` at present. example:
https://github.com/peterbecich/halogen-chess/tree/source-repository-example
My goal is to have the project buildable by either plain Cabal, or Nix with `haskell-flake`.

Can `haskell-flake`'s `cabal.project` parser be modified to ignore the `source-repository-package`?

**Comment by srid** [comment]
haskell-flake itself ignores everything but `packages` -- so it does ignore `source-repository-package`. Maybe `cabal` (as invoked from Nix) is the only one that is forcefully using it?

In your first project, your only local package lives at project root, hence `/cabal.project` gets included in sdist of your only package. You can have haskell-flake ignore cabal.project entirely by manually specifying the path:

```nix
haskellProjects.default = {
  packages.halogen-chess.source = builtins.path { path = ./.; filter = path: _: baseNameOf path != "cabal.project";;
}
```

(Perhaps we should have a `packages.*.filter` option to specify these patterns using https://www.tweag.io/blog/2023-11-28-file-sets/ ?)

**Comment by peterbecich** [comment]
You're correct, thanks! Moving `halogen-chess.cabal` into a subdirectory fixes it: https://github.com/peterbecich/halogen-chess/tree/fix-cabal-project-in-nix/haskell by keeping `cabal.project` out of the sdist.

I can't get the second solution working; both of these cause `infinite recursion`.


```
packages = {
  halogen-chess.source = builtins.path {
    path = ./.;
    # filter = path: _: baseNameOf path != "cabal.project";
  };
};
```

```
packages = {
  halogen-chess.source = builtins.path {
    path = ./.;
    filter = path: _: baseNameOf path != "cabal.project";
  };
};
```

**Comment by peterbecich** [comment]
I'm going to try to reproduce this with the fundamental `make-package-set` functions
https://github.com/NixOS/nixpkgs/blob/e35d582350f5b92c769161532a663fb7fcce74f7/pkgs/development/haskell-modules/make-package-set.nix#L552-L597 , and then file an issue in `nixpkgs`

**Comment by srid** [comment]
It needs to be `self` instead of `./.`; but `builtins.filter` still doesn't work due to the way local packages are detected. This requires some debugging (and I have other things of priority to take care), so in the meanwhile perhaps you want to move `cabal.project` out of the way somehow? Eg. `git mv cabal.project cabal.project.template` (and ask users to move it back when using non-nix setup).

But yea, leveraging the `--offline` flag, while having it work correctly, would be one proper fix here. Another solution is to implement the `filter` option I mentioned above. This seems like a trivial solution that's not hacky.

**Comment by peterbecich** [comment]
Thanks for your time, @srid
